File: certification_server/src/services/authcontrol/inbound_auth_control_svc.go
package authcontrolsvc

import (
	"context"
	"fmt"
	"math"
	"strings"
	"sync"
	"time"

	authcontrolif "certification_server/src/iface/authcontrol"
	authmodel "certification_server/src/models/auth"
	commonmodel "certification_server/src/models/common"
	modelsystem "certification_server/src/models/system"
)

var _ authcontrolif.IInboundAuthControl = (*InboundAuthControlService)(nil)

// InboundAuthControlService 为认证中心提供本地级入站限流能力。
// 该实现不调用其他服务，仅依赖本地配置与进程内计数器。
type InboundAuthControlService struct {
	cfg modelsystem.AuthControlConfig
	now func() time.Time

	mu           sync.Mutex
	fixedWindows map[string]*fixedWindowState
	tokenBuckets map[string]*tokenBucketState
}

type fixedWindowState struct {
	WindowStartedAt time.Time
	WindowResetAt   time.Time
	Hits            int64
}

type tokenBucketState struct {
	Tokens       float64
	LastRefillAt time.Time
	LastSeenAt   time.Time
	Capacity     float64
	RefillPerSec float64
}

// NewInboundAuthControlService 构造认证中心本地限流门面。
func NewInboundAuthControlService(cfg modelsystem.AuthControlConfig) authcontrolif.IInboundAuthControl {
	return &InboundAuthControlService{
		cfg:          cfg,
		now:          time.Now,
		fixedWindows: map[string]*fixedWindowState{},
		tokenBuckets: map[string]*tokenBucketState{},
	}
}

// EnforceInbound 将入站上下文转换为本地限流决策。
func (s *InboundAuthControlService) EnforceInbound(
	ctx context.Context,
	req *authcontrolif.InboundControlRequest,
) (*authcontrolif.InboundControlResult, error) {
	if ctx != nil {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		default:
		}
	}
	if req == nil || req.RateLimitInput == nil {
		return nil, &modelsystem.ErrInboundRateLimitInputNil
	}

	descriptor, err := authcontrolif.Build(req.RateLimitInput)
	if err != nil {
		return nil, err
	}

	now := s.now().UTC()
	decision := s.decide(now, descriptor)
	return &authcontrolif.InboundControlResult{RateLimitDecision: decision}, nil
}

func (s *InboundAuthControlService) decide(now time.Time, descriptor *authmodel.RateLimitDescriptor) *authmodel.RateLimitDecision {
	if descriptor == nil {
		return &authmodel.RateLimitDecision{
			Allowed: false,
			Reason:  "rate limit descriptor is nil",
		}
	}

	if !s.cfg.Enabled {
		return &authmodel.RateLimitDecision{
			Allowed:    true,
			Remaining:  -1,
			SubjectKey: buildSubjectKey(descriptor, s.cfg.Subject),
			Reason:     "auth control disabled",
		}
	}

	rule := s.cfg
	if !ruleMatchesConfig(rule, descriptor) {
		return &authmodel.RateLimitDecision{
			Allowed:    true,
			Remaining:  -1,
			SubjectKey: buildSubjectKey(descriptor, s.cfg.Subject),
			Reason:     "no matching local auth control rule",
		}
	}

	subjectKey := buildSubjectKey(descriptor, rule.Subject)
	if subjectKey == "" {
		subjectKey = buildCompositeSubjectKey(descriptor)
	}
	if subjectKey == "" {
		subjectKey = "anonymous"
	}

	switch rule.Algorithm {
	case authmodel.RateLimitTokenBucket:
		return s.decideTokenBucket(now, rule, descriptor, subjectKey)
	case authmodel.RateLimitSlidingWindow:
		fallthrough
	case authmodel.RateLimitFixedWindow:
		fallthrough
	default:
		return s.decideFixedWindow(now, rule, descriptor, subjectKey)
	}
}

func (s *InboundAuthControlService) decideFixedWindow(
	now time.Time,
	rule modelsystem.AuthControlConfig,
	descriptor *authmodel.RateLimitDescriptor,
	subjectKey string,
) *authmodel.RateLimitDecision {
	windowSec := rule.WindowSec
	if windowSec <= 0 {
		windowSec = 60
	}
	window := time.Duration(windowSec) * time.Second
	if window <= 0 {
		window = 60 * time.Second
	}

	limit := rule.Limit
	if limit <= 0 {
		limit = 600
	}

	key := bucketKeyString(rule.RuleID, descriptor, subjectKey)

	s.mu.Lock()
	state := s.fixedWindows[key]
	if state == nil || !now.Before(state.WindowResetAt) {
		state = &fixedWindowState{
			WindowStartedAt: now,
			WindowResetAt:   now.Add(window),
			Hits:            0,
		}
		s.fixedWindows[key] = state
	}
	state.Hits++
	hits := state.Hits
	windowResetAt := state.WindowResetAt
	s.mu.Unlock()

	remaining := limit - hits
	if remaining < 0 {
		remaining = 0
	}

	if hits > limit {
		retryAfter := int64(math.Ceil(windowResetAt.Sub(now).Seconds()))
		if retryAfter < 1 {
			retryAfter = 1
		}
		return &authmodel.RateLimitDecision{
			Allowed:        false,
			ViolatedRuleID: rule.RuleID,
			RetryAfterSec:  retryAfter,
			Remaining:      0,
			SubjectKey:     subjectKey,
			Reason:         fmt.Sprintf("rate limited by local rule %s", rule.RuleID),
		}
	}

	return &authmodel.RateLimitDecision{
		Allowed:        true,
		ViolatedRuleID: "",
		RetryAfterSec:  0,
		Remaining:      remaining,
		SubjectKey:     subjectKey,
		Reason:         "allowed",
	}
}

func (s *InboundAuthControlService) decideTokenBucket(
	now time.Time,
	rule modelsystem.AuthControlConfig,
	descriptor *authmodel.RateLimitDescriptor,
	subjectKey string,
) *authmodel.RateLimitDecision {
	windowSec := rule.WindowSec
	if windowSec <= 0 {
		windowSec = 60
	}
	limit := rule.Limit
	if limit <= 0 {
		limit = 600
	}
	burst := rule.Burst
	if burst < 0 {
		burst = 0
	}

	capacity := float64(limit + burst)
	if capacity <= 0 {
		capacity = float64(limit)
	}
	if capacity <= 0 {
		capacity = 600
	}
	refillPerSec := float64(limit) / float64(windowSec)
	if refillPerSec <= 0 {
		refillPerSec = 1
	}

	key := bucketKeyString(rule.RuleID, descriptor, subjectKey)

	s.mu.Lock()
	state := s.tokenBuckets[key]
	if state == nil {
		state = &tokenBucketState{
			Tokens:       capacity,
			LastRefillAt: now,
			LastSeenAt:   now,
			Capacity:     capacity,
			RefillPerSec: refillPerSec,
		}
		s.tokenBuckets[key] = state
	}
	elapsed := now.Sub(state.LastRefillAt).Seconds()
	if elapsed > 0 {
		state.Tokens = math.Min(state.Capacity, state.Tokens+elapsed*state.RefillPerSec)
		state.LastRefillAt = now
	}
	state.LastSeenAt = now
	state.Capacity = capacity
	state.RefillPerSec = refillPerSec

	if state.Tokens >= 1 {
		state.Tokens -= 1
		remaining := int64(math.Floor(state.Tokens))
		if remaining < 0 {
			remaining = 0
		}
		s.mu.Unlock()
		return &authmodel.RateLimitDecision{
			Allowed:        true,
			ViolatedRuleID: "",
			RetryAfterSec:  0,
			Remaining:      remaining,
			SubjectKey:     subjectKey,
			Reason:         "allowed",
		}
	}

	missing := 1 - state.Tokens
	retryAfter := int64(math.Ceil(missing / state.RefillPerSec))
	if retryAfter < 1 {
		retryAfter = 1
	}
	state.Tokens = math.Max(0, state.Tokens)
	s.mu.Unlock()

	return &authmodel.RateLimitDecision{
		Allowed:        false,
		ViolatedRuleID: rule.RuleID,
		RetryAfterSec:  retryAfter,
		Remaining:      0,
		SubjectKey:     subjectKey,
		Reason:         fmt.Sprintf("rate limited by local rule %s", rule.RuleID),
	}
}

func ruleMatchesConfig(cfg modelsystem.AuthControlConfig, descriptor *authmodel.RateLimitDescriptor) bool {
	if descriptor == nil {
		return false
	}
	if descriptor.Scope != cfg.Scope {
		return false
	}
	if cfg.MatchModule != "" && !strings.EqualFold(strings.TrimSpace(descriptor.Module), cfg.MatchModule) {
		return false
	}
	if cfg.MatchAction != "" && !strings.EqualFold(strings.TrimSpace(descriptor.Action), cfg.MatchAction) {
		return false
	}
	if cfg.MatchRoute != "" && strings.TrimSpace(descriptor.Route) != cfg.MatchRoute {
		return false
	}
	if len(cfg.MatchMethods) > 0 && !stringListContainsFold(cfg.MatchMethods, descriptor.Method) {
		return false
	}
	if cfg.RequireAuthenticated && !descriptor.Authenticated {
		return false
	}
	if len(cfg.MatchEntityTypes) > 0 && !entityTypeInList(cfg.MatchEntityTypes, descriptor.EntityType) {
		return false
	}
	if len(cfg.MatchTokenTypes) > 0 && !tokenTypeInList(cfg.MatchTokenTypes, descriptor.TokenType) {
		return false
	}
	if len(cfg.MatchGatewayIDs) > 0 && !stringListContainsFold(cfg.MatchGatewayIDs, descriptor.GatewayID) {
		return false
	}
	if len(cfg.MatchSourceServices) > 0 && !stringListContainsFold(cfg.MatchSourceServices, descriptor.SourceService) {
		return false
	}
	if len(cfg.MatchTargetServices) > 0 && !stringListContainsFold(cfg.MatchTargetServices, descriptor.TargetService) {
		return false
	}
	if len(cfg.MatchScopes) > 0 && !stringListContainsAllFold(descriptor.Scopes, cfg.MatchScopes) {
		return false
	}
	if len(cfg.MatchTags) > 0 && !tagMapMatches(descriptor.Tags, cfg.MatchTags) {
		return false
	}
	return true
}

func buildSubjectKey(descriptor *authmodel.RateLimitDescriptor, subject authmodel.RateLimitSubjectType) string {
	if descriptor == nil {
		return ""
	}
	switch subject {
	case authmodel.SubjectComposite:
		return buildCompositeSubjectKey(descriptor)
	default:
		value := strings.TrimSpace(descriptor.SubjectValue(subject))
		if value != "" {
			return value
		}
		return buildCompositeSubjectKey(descriptor)
	}
}

func buildCompositeSubjectKey(descriptor *authmodel.RateLimitDescriptor) string {
	if descriptor == nil {
		return ""
	}
	parts := []string{
		"scope=" + strings.TrimSpace(string(descriptor.Scope)),
		"module=" + strings.TrimSpace(descriptor.Module),
		"action=" + strings.TrimSpace(descriptor.Action),
		"route=" + strings.TrimSpace(descriptor.Route),
		"method=" + strings.TrimSpace(descriptor.Method),
		"source_ip=" + strings.TrimSpace(descriptor.SourceIP),
		"gateway_id=" + strings.TrimSpace(descriptor.GatewayID),
		"client_id=" + strings.TrimSpace(descriptor.ClientID),
		"source_service=" + strings.TrimSpace(descriptor.SourceService),
		"target_service=" + strings.TrimSpace(descriptor.TargetService),
		"entity_type=" + strings.TrimSpace(string(descriptor.EntityType)),
		"entity_id=" + strings.TrimSpace(descriptor.EntityID),
		"principal_id=" + strings.TrimSpace(descriptor.PrincipalID),
		"session_id=" + strings.TrimSpace(descriptor.SessionID),
		"token_id=" + strings.TrimSpace(descriptor.TokenID),
	}
	filtered := make([]string, 0, len(parts))
	for _, item := range parts {
		if strings.HasSuffix(item, "=") {
			continue
		}
		filtered = append(filtered, item)
	}
	if len(filtered) == 0 {
		return ""
	}
	return strings.Join(filtered, "|")
}

func bucketKeyString(ruleID string, descriptor *authmodel.RateLimitDescriptor, subjectKey string) string {
	module := ""
	action := ""
	scope := ""
	if descriptor != nil {
		module = strings.TrimSpace(descriptor.Module)
		action = strings.TrimSpace(descriptor.Action)
		scope = strings.TrimSpace(string(descriptor.Scope))
	}
	return strings.Join([]string{
		strings.TrimSpace(ruleID),
		scope,
		module,
		action,
		strings.TrimSpace(subjectKey),
	}, "|")
}

func stringListContainsFold(items []string, value string) bool {
	needle := strings.TrimSpace(value)
	if needle == "" {
		return false
	}
	for _, item := range items {
		if strings.EqualFold(strings.TrimSpace(item), needle) {
			return true
		}
	}
	return false
}

func stringListContainsAllFold(haystack []string, needles []string) bool {
	if len(needles) == 0 {
		return true
	}
	for _, needle := range needles {
		if !stringListContainsFold(haystack, needle) {
			return false
		}
	}
	return true
}

func entityTypeInList(items []commonmodel.EntityType, value commonmodel.EntityType) bool {
	needle := strings.TrimSpace(strings.ToLower(string(value)))
	if needle == "" {
		return false
	}
	for _, item := range items {
		if strings.EqualFold(strings.TrimSpace(string(item)), needle) {
			return true
		}
	}
	return false
}

func tokenTypeInList(items []authmodel.TokenType, value authmodel.TokenType) bool {
	needle := strings.TrimSpace(strings.ToLower(string(value)))
	if needle == "" {
		return false
	}
	for _, item := range items {
		if strings.EqualFold(strings.TrimSpace(string(item)), needle) {
			return true
		}
	}
	return false
}

func tagMapMatches(actual map[string]string, expected map[string]string) bool {
	if len(expected) == 0 {
		return true
	}
	if len(actual) == 0 {
		return false
	}
	for key, expectedValue := range expected {
		actualValue, ok := actual[key]
		if !ok {
			return false
		}
		if strings.TrimSpace(actualValue) != strings.TrimSpace(expectedValue) {
			return false
		}
	}
	return true
}

File: data_worker/src/services/authcontrol/inbound_auth_control_svc.py
from __future__ import annotations

from dataclasses import dataclass
from collections.abc import Sequence
from math import ceil, floor
from threading import Lock
import time
from collections.abc import Callable

from src.iface.authcontrol.auth_control import (
    Build,
    InboundControlRequest,
    InboundControlResult,
    IInboundAuthControl,
)
from src.models.auth.ratelimit import RateLimitDecision, RateLimitDescriptor
from src.models.sys.config import AuthControlConfig


DEFAULT_AUTH_CONTROL_MODULE = "data_worker"


@dataclass(slots=True)
class _FixedWindowState:
    window_started_at: float
    window_reset_at: float
    hits: int = 0


@dataclass(slots=True)
class _TokenBucketState:
    tokens: float
    last_refill_at: float
    last_seen_at: float
    capacity: float
    refill_per_sec: float


class InboundAuthControlService(IInboundAuthControl):
    """data_worker 的本地入站限流门面。"""

    def __init__(
        self,
        cfg: AuthControlConfig | None = None,
        *,
        now: Callable[[], float] | None = None,
    ) -> None:
        self._cfg = (cfg or AuthControlConfig()).normalized(DEFAULT_AUTH_CONTROL_MODULE)
        self._now = now or time.time
        self._lock = Lock()
        self._fixed_windows: dict[str, _FixedWindowState] = {}
        self._token_buckets: dict[str, _TokenBucketState] = {}

    async def enforce_inbound(self, req: InboundControlRequest) -> InboundControlResult:
        if req is None or req.rate_limit_input is None:
            raise ValueError("inbound rate limit input is nil")

        descriptor = Build(req.rate_limit_input)
        decision = self._decide(self._now(), descriptor)
        return InboundControlResult(rate_limit_decision=decision)

    def _decide(
        self,
        now: float,
        descriptor: RateLimitDescriptor | None,
    ) -> RateLimitDecision:
        if descriptor is None:
            return RateLimitDecision(
                allowed=False,
                violated_rule_id="",
                retry_after_sec=0,
                remaining=0,
                subject_key="",
                reason="rate limit descriptor is nil",
            )

        if not self._cfg.enabled:
            return RateLimitDecision(
                allowed=True,
                violated_rule_id="",
                retry_after_sec=0,
                remaining=-1,
                subject_key=_build_subject_key(descriptor, self._cfg.subject),
                reason="auth control disabled",
            )

        rule = self._cfg
        if not _rule_matches_config(rule, descriptor):
            return RateLimitDecision(
                allowed=True,
                violated_rule_id="",
                retry_after_sec=0,
                remaining=-1,
                subject_key=_build_subject_key(descriptor, self._cfg.subject),
                reason="no matching local auth control rule",
            )

        subject_key = _build_subject_key(descriptor, rule.subject)
        if subject_key == "":
            subject_key = _build_composite_subject_key(descriptor)
        if subject_key == "":
            subject_key = "anonymous"

        match rule.algorithm:
            case "token_bucket":
                return self._decide_token_bucket(now, rule, descriptor, subject_key)
            case "sliding_window":
                return self._decide_fixed_window(now, rule, descriptor, subject_key)
            case _:
                return self._decide_fixed_window(now, rule, descriptor, subject_key)

    def _decide_fixed_window(
        self,
        now: float,
        rule: AuthControlConfig,
        descriptor: RateLimitDescriptor,
        subject_key: str,
    ) -> RateLimitDecision:
        window_sec = rule.window_sec if rule.window_sec > 0 else 60
        limit = rule.limit if rule.limit > 0 else 600

        key = _bucket_key_string(rule.rule_id, descriptor, subject_key)

        with self._lock:
            state = self._fixed_windows.get(key)
            if state is None or now >= state.window_reset_at:
                state = _FixedWindowState(
                    window_started_at=now,
                    window_reset_at=now + window_sec,
                    hits=0,
                )
                self._fixed_windows[key] = state
            state.hits += 1
            hits = state.hits
            window_reset_at = state.window_reset_at

        remaining = limit - hits
        if remaining < 0:
            remaining = 0

        if hits > limit:
            retry_after = int(ceil(window_reset_at - now))
            if retry_after < 1:
                retry_after = 1
            return RateLimitDecision(
                allowed=False,
                violated_rule_id=rule.rule_id,
                retry_after_sec=retry_after,
                remaining=0,
                subject_key=subject_key,
                reason=f"rate limited by local rule {rule.rule_id}",
            )

        return RateLimitDecision(
            allowed=True,
            violated_rule_id="",
            retry_after_sec=0,
            remaining=remaining,
            subject_key=subject_key,
            reason="allowed",
        )

    def _decide_token_bucket(
        self,
        now: float,
        rule: AuthControlConfig,
        descriptor: RateLimitDescriptor,
        subject_key: str,
    ) -> RateLimitDecision:
        window_sec = rule.window_sec if rule.window_sec > 0 else 60
        limit = rule.limit if rule.limit > 0 else 600
        burst = rule.burst if rule.burst >= 0 else 0

        capacity = float(limit + burst)
        if capacity <= 0:
            capacity = float(limit)
        if capacity <= 0:
            capacity = 600.0

        refill_per_sec = float(limit) / float(window_sec)
        if refill_per_sec <= 0:
            refill_per_sec = 1.0

        key = _bucket_key_string(rule.rule_id, descriptor, subject_key)

        with self._lock:
            state = self._token_buckets.get(key)
            if state is None:
                state = _TokenBucketState(
                    tokens=capacity,
                    last_refill_at=now,
                    last_seen_at=now,
                    capacity=capacity,
                    refill_per_sec=refill_per_sec,
                )
                self._token_buckets[key] = state

            elapsed = now - state.last_refill_at
            if elapsed > 0:
                state.tokens = min(state.capacity, state.tokens + elapsed * state.refill_per_sec)
                state.last_refill_at = now
            state.last_seen_at = now
            state.capacity = capacity
            state.refill_per_sec = refill_per_sec

            if state.tokens >= 1:
                state.tokens -= 1
                remaining = int(floor(state.tokens))
                if remaining < 0:
                    remaining = 0
                return RateLimitDecision(
                    allowed=True,
                    violated_rule_id="",
                    retry_after_sec=0,
                    remaining=remaining,
                    subject_key=subject_key,
                    reason="allowed",
                )

            missing = 1 - state.tokens
            retry_after = int(ceil(missing / state.refill_per_sec))
            if retry_after < 1:
                retry_after = 1
            state.tokens = max(0.0, state.tokens)

        return RateLimitDecision(
            allowed=False,
            violated_rule_id=rule.rule_id,
            retry_after_sec=retry_after,
            remaining=0,
            subject_key=subject_key,
            reason=f"rate limited by local rule {rule.rule_id}",
        )


def NewInboundAuthControlService(cfg: AuthControlConfig | None = None) -> IInboundAuthControl:
    return InboundAuthControlService(cfg)


def _rule_matches_config(rule: AuthControlConfig, descriptor: RateLimitDescriptor) -> bool:
    if descriptor.scope != rule.scope:
        return False
    if rule.match_module and descriptor.module.strip().lower() != rule.match_module.lower():
        return False
    if rule.match_action and descriptor.action.strip().lower() != rule.match_action.lower():
        return False
    if rule.match_route and descriptor.route.strip() != rule.match_route:
        return False
    if rule.match_methods and not _string_list_contains_fold(rule.match_methods, descriptor.method):
        return False
    if rule.require_authenticated and not descriptor.authenticated:
        return False
    if rule.match_entity_types and not _entity_type_in_list(rule.match_entity_types, descriptor.entity_type):
        return False
    if rule.match_token_types and not _token_type_in_list(rule.match_token_types, descriptor.token_type):
        return False
    if rule.match_gateway_ids and not _string_list_contains_fold(rule.match_gateway_ids, descriptor.gateway_id):
        return False
    if rule.match_source_services and not _string_list_contains_fold(rule.match_source_services, descriptor.source_service):
        return False
    if rule.match_target_services and not _string_list_contains_fold(rule.match_target_services, descriptor.target_service):
        return False
    if rule.match_scopes and not _string_list_contains_all_fold(descriptor.scopes, rule.match_scopes):
        return False
    if rule.match_tags and not _tag_map_matches(descriptor.tags, rule.match_tags):
        return False
    return True


def _build_subject_key(descriptor: RateLimitDescriptor, subject: str) -> str:
    if subject == "composite":
        return _build_composite_subject_key(descriptor)

    value = descriptor.subject_value(subject)
    if value:
        return value.strip()
    return _build_composite_subject_key(descriptor)


def _build_composite_subject_key(descriptor: RateLimitDescriptor) -> str:
    parts = [
        f"scope={str(descriptor.scope).strip()}",
        f"module={descriptor.module.strip()}",
        f"action={descriptor.action.strip()}",
        f"route={descriptor.route.strip()}",
        f"method={descriptor.method.strip()}",
        f"source_ip={descriptor.source_ip.strip()}",
        f"gateway_id={descriptor.gateway_id.strip()}",
        f"client_id={descriptor.client_id.strip()}",
        f"source_service={descriptor.source_service.strip()}",
        f"target_service={descriptor.target_service.strip()}",
        f"entity_type={str(descriptor.entity_type).strip()}",
        f"entity_id={descriptor.entity_id.strip()}",
        f"principal_id={descriptor.principal_id.strip()}",
        f"session_id={descriptor.session_id.strip()}",
        f"token_id={descriptor.token_id.strip()}",
    ]
    filtered = [item for item in parts if not item.endswith("=")]
    if not filtered:
        return ""
    return "|".join(filtered)


def _bucket_key_string(rule_id: str, descriptor: RateLimitDescriptor, subject_key: str) -> str:
    return "|".join(
        [
            rule_id.strip(),
            str(descriptor.scope).strip(),
            descriptor.module.strip(),
            descriptor.action.strip(),
            subject_key.strip(),
        ]
    )


def _string_list_contains_fold(items: list[str], value: str) -> bool:
    needle = value.strip()
    if needle == "":
        return False
    for item in items:
        if item.strip().lower() == needle.lower():
            return True
    return False


def _string_list_contains_all_fold(haystack: list[str], needles: list[str]) -> bool:
    if not needles:
        return True
    for needle in needles:
        if not _string_list_contains_fold(haystack, needle):
            return False
    return True


def _entity_type_in_list(items: Sequence[object], value: str) -> bool:
    needle = value.strip().lower()
    if needle == "":
        return False
    for item in items:
        if str(item).strip().lower() == needle:
            return True
    return False


def _token_type_in_list(items: Sequence[object], value: str) -> bool:
    needle = value.strip().lower()
    if needle == "":
        return False
    for item in items:
        if str(item).strip().lower() == needle:
            return True
    return False


def _tag_map_matches(actual: dict[str, str], expected: dict[str, str]) -> bool:
    if not expected:
        return True
    if not actual:
        return False
    for key, expected_value in expected.items():
        if key not in actual:
            return False
        if actual[key].strip() != expected_value.strip():
            return False
    return True